plugin/evm/wrapped_block.go

```diff
@@ -387,6 +396,13 @@ func (b *wrappedBlock) syntacticVerify() error {
 		case !headerExtra.BlockGasCost.IsUint64():
 			return fmt.Errorf("too large blockGasCost: %d", headerExtra.BlockGasCost)
 		}
+
+		// In Granite, block gas cost must be exactly 0.
+		if rulesExtra.IsGranite {
+			if !utils.BigEqual(headerExtra.BlockGasCost, common.Big0) {
+				return fmt.Errorf("%w: have %d, expected 0", errNonZeroBlockGasCostGranite, headerExtra.BlockGasCost)
+			}
+		}
```

Comment:
This seems to be duplicate verification from what we are already doing in `consensus.go` why is this added here?

Reply:
My understanding on why this is done is maybe because these paths guard different entry points:

- VM-level (wrapped_block.Verify): rejects incoming/parsing blocks with non‑zero BlockGasCost under Granite, regardless of which consensus engine is used.
- Consensus-level (consensus/dummy.Finalize): re-validates during block finalization/assembly to catch violations in locally built blocks and ensure engine-driven correctness.

Reply:
Block verification flows from:
- VM level through to consensus-level.

There is never a time that one is performed but the other isn't. (Even in build block, we actually fully verify the block before returning it to the consensus engine)